Update uperf metrics for asymmetric workload

rsevilla87 opened the pull request:
We need to update uperf stdout parsing to index the right write/read/nthreads metrics.
 
Signed-off-by: Raul Sevilla <rsevilla@redhat.com>

rsevilla87 commented:
/rerun all

rsevilla87 commented:
Example document:
```json
{
  "_index": "ripsaw-uperf-results",
  "_type": "_doc",
  "_id": "9745004cbac2bc75ea972bcb711fcddbc7a7abbe5dc4126c322c25a195fa659f",
  "_version": 1,
  "_score": null,
  "_source": {
    "workload": "uperf",
    "uuid": "f9ea29f5-7c9d-5f26-b00f-5ad775f50d5a",
    "user": "ripsaw",
    "cluster_name": "myk8scluster",
    "hostnetwork": "False",
    "iteration": 2,
    "remote_ip": "10.129.1.50",
    "client_ips": "10.128.1.20 ",
    "uperf_ts": "2020-12-02T11:47:33.740000",
    "service_ip": "False",
    "bytes": 781352941,
    "norm_byte": 399515629,
    "ops": 895187,
    "norm_ops": 471187,
    "norm_ltcy": 2.973484282613644,
    "kind": "pod",
    "client_node": "",
    "server_node": "",
    "num_pairs": "1",
    "multus_client": "",
    "test": "bidirec",
    "protocol": "udp",
    "write_message_size": 1024,
    "read_message_size": 512,
    "num_threads": 2,
    "results": [
      [
        "1606909651338.8020",
        "0",
        "0"
      ],
      [
        "1606909652339.1553",
        "381837312",
        "424000"
      ],
      [
        "1606909653740.2224",
        "781352941",
        "895187"
      ]
    ]
  },
  "fields": {
    "uperf_ts": [
      "2020-12-02T11:47:33.740Z"
    ]
  },
  "sort": [
    1606909653740
  ]
}
```

smalleni reviewed (approved):
LGTM

jtaleric reviewed (commented):
With this change, the index will change... we are dropping test_type for test, and adding new keys... This will break touchstone[1]... 

There is a bunch of nuance here.. My concern is if we break other tooling... and if we change the structure of the doc, how does that impact our comparison. We would need to code up different paths based on if `test` exists vs `test_type` same thing with the message_size change.

[1] https://github.com/cloud-bulldozer/benchmark-comparison/blob/master/src/touchstone/benchmarks/uperf.py#L55-L57

dustinblack commented:
Thanks for lookin at this @rsevilla87! It was on my to-do list after the benchmark-operator update, but you beat me to it. ;)

Agreeing with @jtaleric, I'm not quite sure what your motivation is in updating the JSON document format beyond the new `wsize` and `rsize` parameters replacing `size`. You could keep the cleaner python and just update the return variables in `_parse_stdout` it seems, but you still have to remember to handle the `len()` for the `results` parameter.

rsevilla87 commented:
> With this change, the index will change... we are dropping test_type for test, and adding new keys... This will break touchstone[1]...
> 
> There is a bunch of nuance here.. My concern is if we break other tooling... and if we change the structure of the doc, how does that impact our comparison. We would need to code up different paths based on if `test` exists vs `test_type` same thing with the message_size change.
> 
> [1] https://github.com/cloud-bulldozer/benchmark-comparison/blob/master/src/touchstone/benchmarks/uperf.py#L55-L57

Good catch, I forgot about test_type and applying len to the results variable. 

On the other hand, the metrics indexed by this benchmark are already broken. Since the asymmetric workload addition, the uperf profile name has now this shape:
`<profile name="{{test}}-{{proto}}-{{wsize}}-{{rsize}}-{{nthr}}">`

And the current implementation to parse this line does the following:
```
        config = re.findall(r"running profile:(.*) \.\.\.", stdout)[0]
        test = re.split("-", config[0])[0]                   
        protocol = re.split("-", config[0])[1]            
        size = re.split("-", config[0])[2]                    <----- This is now using the value of wsize
        nthr = re.split("-", config[0])[3]                    <----- This is now using the vaue of rsize instead nthr
```

rsevilla87 commented:
Document after fixing:
```
{
  "_index": "ripsaw-uperf-results",
  "_type": "_doc",
  "_id": "dfca84e543373cfe8bd599c2c767edfe8faf458a024845f9dbfd57a863e4360a",
  "_version": 1,
  "_score": null,
  "_source": {
    "workload": "uperf",
    "uuid": "6abda416-91ae-5a5c-8aef-6e463defb5db",
    "user": "ripsaw",
    "cluster_name": "myk8scluster",
    "hostnetwork": "False",
    "iteration": 2,
    "remote_ip": "10.129.1.238",
    "client_ips": "10.128.1.51 ",
    "uperf_ts": "2020-12-03T10:33:40.493000",
    "service_ip": "False",
    "bytes": 427319289,
    "norm_byte": 229638137,
    "ops": 834615,
    "norm_ops": 448519,
    "norm_ltcy": 3.128494465082583,
    "kind": "pod",
    "client_node": "",
    "server_node": "",
    "num_pairs": "1",
    "multus_client": "",
    "networkpolicy": "True",
    "test_type": "bidirec",
    "protocol": "udp",
    "write_message_size": 512,
    "read_message_size": 512,
    "num_threads": 2,
    "duration": 3
  },
  "fields": {
    "uperf_ts": [
      "2020-12-03T10:33:40.493Z"
    ]
  },
  "highlight": {
    "uuid.keyword": [
      "@kibana-highlighted-field@6abda416-91ae-5a5c-8aef-6e463defb5db@/kibana-highlighted-field@"
    ],
    "uuid": [
      "@kibana-highlighted-field@6abda416@/kibana-highlighted-field@-@kibana-highlighted-field@91ae@/kibana-highlighted-field@-@kibana-highlighted-field@5a5c@/kibana-highlighted-field@-@kibana-highlighted-field@8aef@/kibana-highlighted-field@-@kibana-highlighted-field@6e463defb5db@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1606991620493
  ]
}```

mohit-sheth reviewed (approved):
LGTM

jtaleric reviewed (approved):
LGTM - there are other scripts that need updating immediately after this change.

jtaleric commented on snafu/uperf_wrapper/trigger_uperf.py:
```diff
@@ -91,18 +87,21 @@ def _run_uperf(self):
+            return results, {"test_type": test_type, "protocol": protocol, "message_size": int(rsize),
+                             "num_threads": int(nthr), "duration": len(results)}
+        else:
+            return results, {"test_type": test_type, "protocol": protocol, "write_message_size": int(wsize),
```
Could we drop `write_` and have `message_size` and `read_message_size`? Can we make the position that `message_size` is write_message_size?

jtaleric reviewed (approved):
Thanks @rsevilla87 I like this new method, it requires little to no change on our existing tooling!
